contracts/Daomon721a.sol

```diff
@@ -5,33 +5,269 @@ import "erc721a-upgradeable/contracts/ERC721AUpgradeable.sol";
+     *      and we can leverage the improvement of the ERC721a standard to save gas.
+     **/
+    uint256 internal deploymentTimeStamp = block.timestamp;
+
+    /**
+     * @notice store all lock infos per token. We use this to avoid storing init all. If not
+     *         set the token is locked with deployment time.
+     * @dev Best use _getLockInfo() to access this mapping. This will take care of minted check
+     *      and will return a default with deployment time if no lock info is set.
+     */
+    mapping(uint256 => LockInfo) internal lockInfo;
+
+    /**
+     * @notice this is the precision of the penalty percentage. So if 10000 is 100% or 200 is 100%
+     */
+    uint32 public constant unlockPenatlyPercentagePrecision = 10_000;
+    // 0 % unlock penalty as default. If set to 0, no penalty will be applied.
+    uint32 unlockPenaltyPercentage = 0;
+
+    // =============================================================
+    //                        Modifiers
+    // =============================================================
+
+    /**
+     * @notice Modifier to check if token is minted.
+     */
+    modifier isMinted(uint256 tokenId) {
+        if (!_exists(tokenId)) revert TokenNotMinted(tokenId);
+        _;
+    }
+
+    /**
+     * @notice Modifier to check if token is locked.
+     */
+    modifier isLocked(uint256 tokenId) {
+        if (!isTokenLocked(tokenId)) revert NotOwnerOfToken(tokenId);
+        _;
+    }
 
-    /* 
-    INITIALIZER
-    */
-    function initialize() public initializerERC721A initializer {
-        __ERC721A_init("Daomon", "DAOMON");
+    /**
+     * @notice Modifier to check if request is coming from a manager address.
+     */
+    modifier onlyManager() {
+        require(
+            hasRole(MANAGER_ROLE, msg.sender),
+            "Caller does not have the required role"
+        );
+        _;
+    }
+
+    // =============================================================
+    //                        CONFIGURATION
+    // =============================================================
+
+    /**
+     * Initialize the contract.
+     *
+     * @param _name The name of the token.
+     * @param _symbol The symbol of the token.
+     * @param _tokenUriProvider The tokenURI provider contract which will create the metadata for a token.
+     * @param _unlockPenaltyPercentage The unlock penalty percentage. 0 = no penalty. 10000 = 100% and the minimal penalty is 1 = 0.01 % penalty.
+     */
+    function initialize(
+        string memory _name,
+        string memory _symbol,
+        ITokenURIProvider _tokenUriProvider,
+        uint32 _unlockPenaltyPercentage
+    ) public initializerERC721A initializer {
+        // set roles
+        _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        _grantRole(MANAGER_ROLE, msg.sender);
+
+        // configure the storage
+        tokenUriProvider = _tokenUriProvider;
+        deploymentTimeStamp = block.timestamp;
```

Comment:
Obwohl forget..

Reply:
If we want this we must do it with a contract. Something like DatetimeProvider because we need this in several places. But I guess this makes no difference as we use block.timestamp; in all places. 

So it just behaves logically the same.